src/platform/bouffalolab/BL702/NetworkCommissioningDriver.cpp

```diff
@@ -229,81 +229,102 @@ void BLWiFiDriver::OnScanWiFiNetworkDone(void * opaque)
-        {
-            ap_num = 1;
-            break;
-        }
-    }
+    ChipLogProgress(DeviceLayer, "expected ssid %s. get %d in total, %d", mScanSSID, pmsg->num, mScanSpecific);
 
-    if (0 == pmsg->num || (mScanSpecific && 0 == ap_num))
+    if (mScanSpecific)
     {
-        ChipLogProgress(DeviceLayer, "No AP found");
-        if (mpScanCallback != nullptr)
+        for (i = 0; i < pmsg->num; i++)
         {
-            mpScanCallback->OnFinished(Status::kSuccess, CharSpan(), nullptr);
-            mpScanCallback = nullptr;
+            if (mScanSpecific && !strcmp(mScanSSID, (char *) (pmsg->records[i].ssid)))
+            {
+                ap_num = 1;
+                break;
+            }
         }
-        return;
-    }
-
-    if (ap_num)
-    {
-        p = pScanResponse = (WiFiScanResponse *) malloc(sizeof(WiFiScanResponse) * ap_num);
     }
     else
     {
-        p = pScanResponse = (WiFiScanResponse *) malloc(sizeof(WiFiScanResponse) * pmsg->num);
-        ap_num            = pmsg->num;
+        ap_num = pmsg->num;
     }
-    for (i = 0; i < pmsg->num; i++)
+
+    if (ap_num)
     {
-        if (mScanSpecific && strcmp(mScanSSID, (char *) (pmsg->records[i].ssid)))
+        p = mScanResponse = (WiFiScanResponse *) malloc(sizeof(WiFiScanResponse) * ap_num);
+        if (mScanResponse == nullptr)
         {
-            continue;
+            return;
         }
 
-        p->security.SetRaw(pmsg->records[i].auth_mode);
-        p->ssidLen  = strlen((char *) pmsg->records[i].ssid) < chip::DeviceLayer::Internal::kMaxWiFiSSIDLength
-             ? strlen((char *) pmsg->records[i].ssid)
-             : chip::DeviceLayer::Internal::kMaxWiFiSSIDLength;
-        p->channel  = pmsg->records[i].channel;
-        p->wiFiBand = chip::DeviceLayer::NetworkCommissioning::WiFiBand::k2g4;
-        p->rssi     = pmsg->records[i].rssi;
-        memcpy(p->ssid, pmsg->records[i].ssid, p->ssidLen);
-        memcpy(p->bssid, pmsg->records[i].bssid, 6);
-
-        if (mScanSpecific)
+        for (i = 0; i < pmsg->num; i++)
         {
-            break;
-        }
-
-        p++;
-    }
-
-    if (CHIP_NO_ERROR == DeviceLayer::SystemLayer().ScheduleLambda([ap_num, pScanResponse]() {
-            BLScanResponseIterator iter(ap_num, pScanResponse);
-            if (GetInstance().mpScanCallback)
+            if (mScanSpecific && strcmp(mScanSSID, (char *) (pmsg->records[i].ssid)))
             {
-                GetInstance().mpScanCallback->OnFinished(Status::kSuccess, CharSpan(), &iter);
-                GetInstance().mpScanCallback = nullptr;
+                continue;
             }
-            else
+
+            p->security.SetRaw(pmsg->records[i].auth_mode);
+            p->ssidLen  = strlen((char *) pmsg->records[i].ssid) < chip::DeviceLayer::Internal::kMaxWiFiSSIDLength
+                 ? strlen((char *) pmsg->records[i].ssid)
+                 : chip::DeviceLayer::Internal::kMaxWiFiSSIDLength;
```

Comment:
consider using bytespan or chip's CopyString 

You can avoid calling strlen((char *) pmsg->records[i].ssid) twice with a local variable

Reply:
copy dest `ssid`  is 32 byte length without null terminator. use `strncpy` to copy ssid to avoid calling strlen twice